x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,203 @@
+            return; // system index primary shard does not exist or is not assigned to this data node
+        }
+        if (event.indicesDeleted().isEmpty() == false) {
+            threadPool.generic().execute(new MaintenanceTask(event));
+        }
+    }
+
+    @Nullable
+    private ShardRouting systemIndexPrimaryShard(final ClusterState state) {
+        final IndexMetadata indexMetadata = state.metadata().index(systemIndexName);
+        if (indexMetadata != null) {
+            final IndexRoutingTable indexRoutingTable = state.routingTable().index(indexMetadata.getIndex());
+            if (indexRoutingTable != null) {
+                return indexRoutingTable.shard(0).primaryShard();
+            }
+        }
+        return null;
+    }
+
+    private static boolean hasSearchableSnapshotWith(final ClusterState state, final SnapshotId snapshotId, final IndexId indexId) {
+        for (IndexMetadata indexMetadata : state.metadata()) {
+            final Settings indexSettings = indexMetadata.getSettings();
+            if (SearchableSnapshotsSettings.isSearchableSnapshotStore(indexSettings)) {
+                final SnapshotId otherSnapshotId = new SnapshotId(
+                    SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSettings),
+                    SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings)
+                );
+                if (Objects.equals(snapshotId, otherSnapshotId)) {
+                    final IndexId otherIndexId = new IndexId(
+                        SNAPSHOT_INDEX_NAME_SETTING.get(indexSettings),
+                        SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                    );
+                    if (Objects.equals(indexId, otherIndexId)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    static QueryBuilder buildDeleteByQuery(int numberOfShards, String snapshotUuid, String indexUuid) {
+        final Set<String> paths = new HashSet<>(numberOfShards);
+        for (int shard = 0; shard < numberOfShards; shard++) {
+            paths.add(String.join("/", snapshotUuid, indexUuid, String.valueOf(shard)));
+        }
+        return QueryBuilders.termsQuery("blob.path", paths);
+    }
+
+    private class MaintenanceTask extends AbstractRunnable {
+
+        private final ClusterChangedEvent event;
+
+        MaintenanceTask(ClusterChangedEvent event) {
+            assert event.indicesDeleted().isEmpty() == false;
+            this.event = Objects.requireNonNull(event);
+        }
+
+        @Override
+        protected void doRun() {
+            final ClusterState state = event.state();
+            for (Index deletedIndex : event.indicesDeleted()) {
+                final IndexMetadata indexMetadata = event.previousState().metadata().index(deletedIndex);
+                if (indexMetadata != null) {
```

Comment:
I think we expect the metadata to exist in the previous state? I think the `if` is fine, but could be accompanied by `assert indexMetadata != null`?

Reply:
Sure